src/rust/storage/seg/src/ttl_buckets/ttl_bucket.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+//! TTL bucket containing a segment chain which stores items with a similar TTL
+//! in an ordered fashion.
+//!
+//! TTL Bucket:
+//! ```text
+//! ┌──────────────┬──────────────┬─────────────┬──────────────┐
+//! │   HEAD SEG   │   TAIL SEG   │     TTL     │   SEGMENTS   │
+//! │              │              │             │              │
+//! │    32 bit    │    32 bit    │    32 bit   │    32 bit    │
+//! ├──────────────┼──────────────┴─────────────┴──────────────┤
+//! │  NEXT MERGE  │                  PADDING                  │
+//! │              │                                           │
+//! │    32 bit    │                  96 bit                   │
+//! ├──────────────┴───────────────────────────────────────────┤
+//! │                         PADDING                          │
+//! │                                                          │
+//! │                         128 bit                          │
+//! ├──────────────────────────────────────────────────────────┤
+//! │                         PADDING                          │
+//! │                                                          │
+//! │                         128 bit                          │
+//! └──────────────────────────────────────────────────────────┘
+//! ```
+
+use crate::*;
+use core::num::NonZeroU32;
+
+/// Each ttl bucket contains a segment chain to store items with a similar TTL
+/// in an ordered fashion. The first segment to expire will be the head of the
+/// segment chain. This allows us to efficiently scan across the [`TtlBuckets`]
+/// and expire segments in an eager fashion.
+pub struct TtlBucket {
+    head: Option<NonZeroU32>,
+    tail: Option<NonZeroU32>,
+    ttl: i32,
+    segments: i32,
```

Comment:
nit: Should this be named `nseg` to be consistent with `Eviction`?

Reply:
Fixed.